editor/src/messages/portfolio/document/node_graph/node_graph_message_handler.rs

```diff
@@ -381,12 +381,30 @@ impl NodeGraphMessageHandler {
 			warn!("Deleting the output node!");
 			return false;
 		}
+
+		let mut first_input_node: Option<NodeInput> = None;
+
+		if !disconnect_nodes {
+			// check whether the deleting node's first (primary) input is a node
+			for (node_id, node) in network.nodes.iter() {
+				if *node_id == deleting_node_id {
+					if node.inputs.len() > 0 {
+						if let NodeInput::Node { .. } = node.inputs[0] {
+							first_input_node = Some(node.inputs[0].clone());
+						}
+					}
+				}
+			}
+		}
+
 		for (node_id, node) in network.nodes.iter_mut() {
 			if *node_id == deleting_node_id {
 				continue;
 			}
 			for (input_index, input) in node.inputs.iter_mut().enumerate() {
-				let NodeInput::Node { node_id, .. } = input else {
+				// skip iteration if not a node, otherwise extract its values for use
+				// node_id here shadows the node_id in the outer for-loop
```

Comment:
The shadowing here is a bit confusing. To fix this, you could use `let NodeInput::Node { node_id: upstream_node_id, output_index, .. }`. Then you probably don't need this comment (as it is pretty clear that we are skipping if it is not a node input).